added type discrimination to public Field type in apis

WesleyClements opened the pull request:
Added a new type for Fields that discriminates the corresponding field properties based on the type property.
I also added a template string literal for the string field type that in practice just makes sure you don't forget the colon (it could be more useful if I didn't make it allow custom field types)
I changed all public functions to take this type so that the editor provides a bit more information about what fields have what properties.
All option variation are based on the docs (which seem to need updating) plus any properties not listed but necessary to appease the type checker.
Just a disclaimer that I did not exhaustively investigate the codebase to verify that all properties used internally are included on their corresponding field type.

oscarotero commented:
Hi. Thanks for contributing with Lume CMS!

I need some time to understand all these changes, because it's too complex for my understanding of TypeScript.

For example: What `Prettify` is for?

WesleyClements commented:
It's not necessary at all for the type discrimination. It's a trick to make the typescript engine actually display the properties a type has in tooltips instead of just a bunch of obtuse type names. Sorry for the complexity. I could probably simplify it some if I took another pass of it.

WesleyClements commented:
I know this change isn't strictly necessary. I just found it mildly frustrating that my code editor was suggesting properties that don't exist on a given field Like option on a choose-list. I kept having to double check the docs to make sure I was using the right property. This change to the type makes it so the editor will only suggest properties that exist for the type you defined (unless you're using a custom field in which case it should be exactly the same as before). To clarify what I said before about not exhaustively investigating, I'm not certain that `UniqueFieldOptionMap` and `FieldOptionPropertySelectionMap` contain all the exceptions to the norm that are implemented internally.

WesleyClements commented:
I have added some comments, revised the names of types, and simplified this or at least made if more like it was. Instead of both adding to and removing from the common options I changed it to only filter the properties for each field type. The functionality should be exactly the same but I hope this is easier to follow.

oscarotero commented on types.ts:
```diff
@@ -84,24 +154,94 @@ export interface Field {
+/**
+ * Matches a string of form `/^.*:\s?.*!?$/` where the first part is the field name and the second part is the field type.
+ */
+export type FieldString = `${string}:${"" | " "}${Field["type"]}${"" | "!"}`;
```
I didn't know Typescript could check string formats, this is great!

My only concern is if the tooltip text is helpful or not. This is what I see:

```
Type '"site textarea"' is not assignable to type 'Field | `${string}:number` | `${string}:number!` | `${string}:object` | `${string}:object!` | `${string}:text` | `${string}:text!` | `${string}:textarea` | `${string}:textarea!` | `${string}:markdown` | `${string}:markdown!` | `${string}:code` | `${string}:code!` | `${string}:datetime` | `${string}:datetime!` | `${s...'.deno-ts(2322)
```

Due it accepts too many options (and even field types not included by default), I'm not sure if this is good idea. Maybe this could be enough?

```ts
export type FieldString = `${string}:${string}${"" | "!"}`;
```

WesleyClements replied:
Yeah that's reasonable. I had hoped to generate the `FieldKey` type based on what strings are passed into the `cms.field` function so that there wouldn't be a need to accommodate custom fields with the very general `string` type. I might take a crack at that next if I have time. I'm not sure that it's particularly useful if type can be any string. Also the autocompletion probably would be helped with a redesign of the api. Like I could image using tagged template strings to generate the field objects in place like `type\`name ("" | "!")\`` (ex. `text\`title!\``) or something, or using a tuple of the form `[name, type ("" | "!")]` (ex. `["title","text!"]`). Not saying that either of those changes are even slightly needed but just putting my thoughts down on paper.

oscarotero replied:
I see. The reason of using strings `"title: text!"` is for brevity and ergonomic. Tagged template strings is an interesting idea, but it would require to import the template function so we lost ergonomic. The use of arrays is also interesting but not so ergonomic as a simple string.

Maybe, people wanting for strict types should just use the object signature.

oscarotero commented on types.ts:
```diff
@@ -84,24 +154,94 @@ export interface Field {
+  upload?: string | false;
+  publicPath?: string;
+  options?: Option[];
+  fields?: Field[];
```
This should be:

```suggestion
  fields?: (Field | FieldString)[];
```

oscarotero reviewed (changes requested):
Hi. I did a quick review, and found a couple of issues.
But I'm not sure if this works fine, because I continue seeing invalid properties. For example "text" field doesn't need the "upload" option, but VSCode autocompletes it and consider it valid:
<img width="693" alt="imaxe" src="https://github.com/user-attachments/assets/336ed074-2bee-4807-9bb6-48aa51e5102a" />

If the problem is the autocomplete, maybe it's easier to create a group of interface and combine them. For example:

```ts
// Basic to all fields
interface BasicProperties {
  name: string;
  type: string;
}

// Common to all fields of type <input ...>
interface InputProperties extends BasicProperties {
  label?: string;
  description?: string;
  view?: string;
  attributes?: {...}
}

// Common to all fields of type <input ...> accepting text values
interface TextField extends InputProperties {
  name: "text" | "textarea" | "email" | "url";
  value: string;
}

// Figure out a way to overload the interface with all possible interfaces
export type DefaultFields = TextField | DateField | NumberField | ...
```

Maybe I'm oversimplifying it, I don't know. I'm just looking for a way to make it more mantainable in the long term. What do you think?

Another missing thing is the description for each property but this is a different issue.

WesleyClements commented:
The auto completion doesn't work in the cms repo project itself but it's seemingly working in my project that imports my fork without changes. I deleted the deno cache and reinstalled it to make sure. I'm a little confused how that is possible. I really only tested in the repo I was importing the cms into so I'm a bit blindsided by this. I can get it to work in the cms repo if I get rid of the union with the `CustomType`. What I think is happening is that the typescript engine is interpreting the `{ type: "text", }` as a `CustomType` so it has all the properties of that a custom type would have. I could be wrong (knock on wood) but I'm pretty certain that the type differentiation is working besides the custom type. The change you suggested (using interface inheritance) would definitely work but I'm fairly certain it wouldn't solve the issue of type have the option of being a generic `string` for the custom fields. I can definitely refactor this to be implemented as you're suggesting if that way would be more comfortable for you. I definitely is important to me that this change doesn't make maintaining this more difficult. I'll work on getting it to work in the cms repo in the mean time.

WesleyClements commented:
I have figured out an intermediary step to getting type inference working. It now works for all built in types but in order for custom types to not cause a type error you have to add them as a generic type argument.
```typescript
cms.collection({
  name: "pages",
  src: "src:pages",
  fields: [{
    name: "title",
    label: "Title",
    type: "text",
    upload: "default", // <- throws error for unknown property
  }],
});
```
But so does
```typescript
cms.collection({
  name: "pages",
  src: "src:pages",
  fields: [{
    name: "custom",
    type: "custom", // <- throws error for unknown type
  }],
});
```
Unless you do
```
cms.collection<"custom>({
  name: "pages",
  src: "src:pages",
  fields: [{
    name: "custom",
    type: "custom",
  }],
});
```
This obviously is a breaking change so not ideal. I'll work to see if I can get the custom field type to be generated as `cms.field` is called. At the very least it should be working basically as it was if you aren't using custom fields.

WesleyClements commented:
So I've made it so that calling `cms.field` returns the modified type. It works great especially for the default fields. I am no longer telling the cms what the default fields are. I'm just adding them using `cms.field` and it adds the type field. The main problem with my solution is that it doesn't retrospectively update the type (not sure that is possible). This means that you have to define fields on the same line as defining the variable or else the type change won't be saved:
```typescript
const cms = lumeCms().field("example", {...}).field("example2", {...});
```
This isn't great as it's a breaking change that prevents plugins from defining fields and just generally is too restrictive. I'll keep working to try and make it more flexible when I can.

WesleyClements commented:
Alright I got it working how I had originally wanted it working. There is now discrimination for the properties of different field types including unknown custom fields.
The change ended up being a relatively small step away from what I had originally so I went ahead and removed all the extra bells and whistles I had implemented. I will keep messing around with implementing more strict typechecking and will make another PR if I figure it out. There shouldn't be any changes to logic or functionality, though I did refactor some logic code to appease the type checker. Also I changed the addition of the core fields into a plugin (this is the only change towards the bigger additions that I left in). I'm more than willing to answer any questions about anything and will gladly refactor anything that isn't super clear.

oscarotero commented:
Hi. I just tried it and the type discrimination seems to work but not for new field types:
<img width="543" alt="imaxe" src="https://github.com/user-attachments/assets/75eef802-fd73-4e79-9a94-1d6aa30a6d56" />
```
Type '"foo"' is not assignable to type '"number" | "object" | "url" | "text" | "textarea" | "markdown" | "code" | "datetime" | "current-datetime" | "date" | "time" | "hidden" | "color" | "email" | "select" | "radio" | ... 4 more ... | "file"'.deno-ts(2322)
```

WesleyClements commented:
Sorry that I didn't test this more thoroughly. I was so excited that I thought I figured it out that I rushed through. It now works as I originally intended. I'm almost 1000% certain. That being said. There is a problem with nested fields. If you use a custom type that is not declared in the outer list it will give an error.
```typescript
cms.collection({
  ...
  fields: [
    {
      ...
      type: "hello", // <- should be fine
     ...
   },
   {
     ...
     type: "object",
     fields: [
       {
         ...
         type: "hi", // <- unknown type
         ...
       },
     ],
     ...
   }
  ]
  ...
})
```
I'll work to resolve this though for the mean time you can declare custom types while creating the Cms object like `lumeCms<"hi">()`

WesleyClements commented:
Alright, I almost certain that I have resolved this. The nested fields no longer error if the field type that wasn't used in the base field array. There won't be autocompletions for those types unless you add it as a generic argument when creating the Cms object. Sorry that I ended up testing so many version that I could have easily known weren't working. I really severely underestimated how ruthless the inclusion of the string type ended up being. Still willing to go over/refactor any of this that is confusing.

oscarotero commented:
Sorry for the late response and thanks for the hard work. I'll need some time to understand all these changes. As said, I'm not an advanced Typescript developer so there are a lot of things here that I don't fully understand, and I'm afraid I won't be able to maintain or extend with new fields. There are types with the purpose to create other types, etc.

I will give a try but gime me some days because I have a lot on my plate right now :)

WesleyClements commented:
I'm working on reworking this so that calling the `cms.field` function allows you to specify the properties of the field which will simplify the typescript significantly. more specifically it would mean that maintenance would just be adding another `cms.field` call and specifying the types. Really I'm only having a problem with allowing unknown field types again. It would be made simpler if I could just force the dev to only use fields that they have added with `cms.field`. I believe this is the intended behavior and that the use of `string` as a possible field type is to accommodate those added fields with simple typescript?

oscarotero commented:
`cms.field` is used [to register all fields](https://lume.land/cms/configuration/custom_fields/). It's not possible to use a field not registered previously with that function. So if it's possible to pass the types to the function, (maybe with a generic like `cms.field<Fieldtype>()`, I don't know), it would be nice.

As you said, the use of string as a field type is to allow other field types that the user can register (I didn't know it was possible to create a type containing the names of all registered fields, default and custom).

WesleyClements commented:
I think I figured it out. It now works that when you call `cms.field` and the returned Cms object will include that new field. You can specify the properties as a generic like `cms.field<{ type: "custom-field"; name: string; }>("custom-field", {...})`. I also made it so that plugins can update the type by returning the Cms object. This is so that field types added as part of a plugin won't be lost. There is no functional difference, it's only for the types. I think this is working in generally the best way possible.
That being said, if any fields or plugins that create fields are not included on the same line as you declare the variable containing your Cms instance the field types will be lost. Because of this, if I had removed support for unknown field types, it would be a breaking change for anyone who wrote code like
```typescript
const cms = lumeCms();
cms.field("custom-field", {...});
```
Luckily, as part of the changes I made for this, I was able to support arbitrary field types without the need for the `string` catch all type. This means that the type discrimination should be working without any breaking changes.
It might eventually be worth reconsidering whether forcing users to call `cms.field` on the same line would be better for type-safety, but for now I think it is may be a bit more ergonomic to allow any type but only do type-checking on defined field types. It will only autocomplete fields that have been added through `cms.field` and when you run the program it will throw an error if the field hasn't been defined.
If you ever decided that type-safety is a priority, it would be very little work to force the user to call `cms.field` on the same line. But, since that would be a breaking change anyways, you might as well rethink a more drastic change that might be more elegant. Specifically, because both
```typescript
const cms = lumeCms().field("custom-field", {...});
```
and
```typescript
const cms = lumeCms();
cms.field("custom-field", {...});
```
are functionally the same but only the first will update the types, users might be rightfully confused by the second not doing what they expect. I don't know what the most elegant solution would be but I could think on it if necessary.
From a functionality perspective, I think this is almost done but I think there are some performance problems with the types. If you could verify that for me it'd be awesome. I'm finding that it feels like it takes a long time to get autocompletion suggestions. It could be something unrelated to the code but, if it isn't, I will need to work to address that before I am satisfied with this.

oscarotero commented:
Thanks!
I think holding the field types in the `cms` instance is problematic as you see with the different behavior of `lumecms().field()` vs `const cms = lumecms(); cms.field()`.

I'd rather to store the types in a global namespace. This is what Lume does for plugins and it works pretty well ([Example](https://github.com/lumeland/lume/blob/main/plugins/og_images.ts#L164))

What do you think about something like this?

```js
declare global {
  namespace LumeCMS {
    export interface Fields {
      text: {} // Properties for "text" field
      file: {} // Properties for "file" field
      // etc
    }
  }
}
```

Any other custom field just need to extend this interface.
Then, the types for autocompletion just have to use the `LumeCMS.Fields` interface to get all possible types.

```js
// From https://stackoverflow.com/questions/49285864/is-there-a-valueof-similar-to-keyof-in-typescript
type ValueOf<T> = T[keyof T];

type FieldType = keyof LumeCMS.Fields;
type FieldsString = `${string}:${"" | " "}${FieldType}${"" | "!"}`;
type FieldsArray = ValueOf<LumeCMS.Fields>;
```

I didn't test it, so I'm not sure if this will work. I'm just trying to propose a more simple approach.

WesleyClements commented:
I went ahead and changed it as you asked. I definitely agree that it would be great to be able to simplify this more. That being said, here are a few big problems that I ran into that lead to an increase in complexity. First is that when using `string & Record<never, never>` as a possible field type it resulted in the fields property here
```typescript
cms.collection({
  ...
  fields: [
    {
      type: "text",
      ...
    },
    {
      type: "hello",
      ...
    },
  ],
})
```
having a type of `Field<string>[]`. This meant that if you ever used a non core field type you would lose all discrimination for the other fields. I resolved this by making `cms.collection` generic which allowed inferring the type `"hello"` and ending up with the type `Field<"text" | "hello">[]`. The second problem that I ran into was that we don't want `Field<"text" | "hello">[]` we want `(Field<"text"> | Field<"hello">)[]`. `Field<"text" | "hello">[]` is effectively `{ type: "text" | "hello" }[]`  when we need `({ type: "text"} | { type: "hello" })[]` for discrimination. That is why `FieldArray` looks the way it does. The last big problem was the type of nested `fields` properties. If you try to define the type of `field` statically, there is no way to accommodate custom fields without using `string & Record<never, never>` which causes the previous problems. To remedy this, we have to make `Field` generic and take in the explicitly defined types and then add the `fields` property on when needed. It's not elegant nor simple but there is no way to do it while still having discrimination. Basically, using `string & Record<never, never>` anywhere prevents consistent discrimination and thus the code must be more complicated to accommodate not using it.

WesleyClements commented:
I should say also that I could simplify some of this if it is okay for these changes to be breaking (like only allowing fields explicitly defined). As far as I can determine, none of the changes I have made will cause errors for any existing code bases.

oscarotero commented:
Hi @WesleyClements
Sorry for the late response.

I'm not sure to understand completely the problem, so I've created a different repository to test the types. I think it's easier to have a place to test different ideas and experimentation.

Take a look to this: https://github.com/lumeland/cms-typo-test

In the `types.ts` file, I have defined different types. For now, all of them extend the basic `Field` interface (this part can be improved to avoid repetition).

Then, I declared the `Lume` global namespace. Inside there's an interface with all declared Fields, and with this we can create the string versions (`StringField`) and a generic `Field` type for type discrimination.

Note that fields containing other fields (like `group`) use the type `Lume.Field[]`.

If you see the `custom-type.ts` file, I've declared the interface for a new field, and added it to the global `Lume.Fields` interface. So it's automatically available everywhere (even as a subfield of the `group` field.

You can open `mod.ts` to play with this prototype and see how the types are correctly checked and autocompleted by the IDE. 

Do you think an approach like this could work? This would make the addition of new fields to LumeCMS easy because you only have to enhance the `Lume.Fields` interface adding the new type.

WesleyClements commented:
I went ahead and interpreted your experimenting as being up for making breaking changes. I revised this so that it is based on your suggestions. My implementation is a bit more complicated as it does more than just give autocompletions for the field properties. You should also now get auto completion for properties inside of the `init` and `transform` functions. This did end up being very similar to what I had tried originally (besides using the namespace and without interfaces) but I think it didn't work because of the requirement to allow `string & Record<never, never>` as a field type. I'm glad you prototyped it as I was feeling a bit pessimistic about finding a simpler solution while not making breaking changes

oscarotero commented:
Ok, I have tested these changes and now it works fine.

My only comment is the missing "options" property for "text" fields (Some text fields use the "options" property to [create a `<datalist>` element](https://github.com/lumeland/cms/blob/main/static/components/f-text.js#L51) with suggestions.

Let me know when you are happy with the changes so I can merge it.
After merging, probably I will make some changes (I'm thinking of moving every field to its own file (`fields/text.ts`, `fields/select.ts`, etc) to make the code clearer and easier to follow. But that's a different change.

WesleyClements commented:
@oscarotero I fixed the `text` field missing options and did some refactoring to set things up so that you might be more comfortable making changes. Emphasis on the hopefully. Should be fine to merge

oscarotero commented:
THANKS so much for all the great work and your patience after 1 month and half!!
As said, probably I'll make some changes, but your contribution was key to have better types.
❤️ ❤️ ❤️ ❤️
